File: src/platforms/android/app/src/main/java/com/WangWangPhone/ui/WeatherRealtimeMemoryCache.kt
package com.WangWangPhone.ui

import java.util.Locale

private const val WEATHER_MEMORY_CACHE_DEFAULT_MAX_AGE_MS = 2 * 60 * 60 * 1000L

object WeatherRealtimeMemoryCache {
    private data class Snapshot(
        val cityKey: String,
        val payload: String,
        val savedAtMillis: Long
    )

    private val lock = Any()
    private var snapshot: Snapshot? = null

    fun save(city: String, payload: String) {
        val normalizedCity = normalizeCity(city)
        if (normalizedCity.isEmpty() || payload.isBlank()) return
        synchronized(lock) {
            snapshot = Snapshot(
                cityKey = normalizedCity,
                payload = payload,
                savedAtMillis = System.currentTimeMillis()
            )
        }
    }

    fun load(city: String, maxAgeMillis: Long = WEATHER_MEMORY_CACHE_DEFAULT_MAX_AGE_MS): String? {
        val normalizedCity = normalizeCity(city)
        if (normalizedCity.isEmpty()) return null
        synchronized(lock) {
            val cached = snapshot ?: return null
            if (cached.cityKey != normalizedCity) return null
            if (System.currentTimeMillis() - cached.savedAtMillis > maxAgeMillis) return null
            return cached.payload
        }
    }

    private fun normalizeCity(city: String): String {
        return city.trim().lowercase(Locale.ROOT)
    }
}

File: src/platforms/ios/WangWangPhone/Core/WeatherRealtimeMemoryCache.swift
import Foundation

enum WeatherRealtimeMemoryCache {
    private struct Snapshot {
        let cityKey: String
        let payload: String
        let savedAt: Date
    }

    private static let lock = NSLock()
    private static var snapshot: Snapshot?
    private static let defaultMaxAge: TimeInterval = 2 * 60 * 60

    static func save(city: String, payload: String) {
        let key = normalizedCity(city)
        guard !key.isEmpty else { return }
        guard !payload.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }

        lock.lock()
        snapshot = Snapshot(cityKey: key, payload: payload, savedAt: Date())
        lock.unlock()
    }

    static func load(city: String, maxAge: TimeInterval = defaultMaxAge) -> String? {
        let key = normalizedCity(city)
        guard !key.isEmpty else { return nil }

        lock.lock()
        defer { lock.unlock() }

        guard let cached = snapshot else { return nil }
        guard cached.cityKey == key else { return nil }
        guard Date().timeIntervalSince(cached.savedAt) <= maxAge else { return nil }
        return cached.payload
    }

    private static func normalizedCity(_ city: String) -> String {
        city.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    }
}
